pkg/storage/stores/shipper/indexshipper/tsdb/compactor_fd_leak_test.go

```diff
@@ -0,0 +1,242 @@
+				if count > maxFDCount {
+					maxFDCount = count
+				}
+				maxFDCountMutex.Unlock()
+				time.Sleep(5 * time.Millisecond) // Check every 5ms for better detection
+			}
+		}
+	}()
+
+	// Call setupBuilder in a goroutine so we can monitor during execution
+	ctx := context.Background()
+	var builder *Builder
+	var builderErr error
+	go func() {
+		close(started)
+		builder, builderErr = setupBuilder(ctx, indexFormat, "user1", idxSet, []Index{})
+		close(done)
+	}()
+
+	// Wait for completion
+	<-done
+	<-monitorDone
+
+	require.NoError(t, builderErr)
+	require.NotNil(t, builder)
+
+	// Get final file descriptor count
+	finalFDCount := getFDCount(t)
+
+	maxFDCountMutex.Lock()
+	peakFDCount := maxFDCount
+	maxFDCountMutex.Unlock()
+
+	// Calculate the increase in file descriptors
+	// With the bug: file descriptors accumulate during the loop (peak will be high)
+	// With the fix: file descriptors are closed immediately (peak should be low)
+	fdIncrease := peakFDCount - initialFDCount
+
+	// Analyze FD samples to see if there's a pattern of accumulation
+	var avgFDCount int
+	if len(fdSamples) > 0 {
+		sum := 0
+		for _, sample := range fdSamples {
+			sum += sample
+		}
+		avgFDCount = sum / len(fdSamples)
+	}
+
+	t.Logf("File descriptor stats: initial=%d, peak=%d, avg=%d, final=%d, increase=%d, numFiles=%d, samples=%d",
+		initialFDCount, peakFDCount, avgFDCount, finalFDCount, fdIncrease, numFiles, len(fdSamples))
+
+	// The key test: with the bug, file descriptors accumulate during the loop.
+	// Each file opened adds ~1-2 FDs (one for the file, possibly one for mmap).
+	// With 30 files, we'd expect to see at least 20+ FDs accumulated if the bug exists.
+	// With the fix, files are closed immediately, so peak should be much lower.
+
+	// Calculate expected accumulation: each file might use 1-2 FDs
+	// With the bug, we'd see accumulation proportional to numFiles
+	// With the fix, we should see a constant overhead (maybe 5-10 FDs)
+	expectedLeakFDs := numFiles * 1 // At least 1 FD per file if leak exists
+	maxAllowedIncrease := 20        // Allow some overhead, but not proportional to numFiles
+
+	if fdIncrease >= expectedLeakFDs/2 {
+		// If we see accumulation proportional to numFiles, it's a leak
+		t.Errorf("File descriptor leak detected! Peak increase: %d (expected leak: ~%d, allowed: %d). "+
+			"This indicates files are not being closed immediately during processing. "+
+			"With the fix, file descriptors should be closed right after use, not deferred until function return. "+
+			"FD samples show accumulation pattern.",
+			fdIncrease, expectedLeakFDs, maxAllowedIncrease)
+	}
+
+	// Verify the builder was populated correctly
+	require.NotNil(t, builder.streams)
+	require.Greater(t, len(builder.streams), 0, "Builder should have streams")
+}
+
+// getFDCount returns the current number of open file descriptors
+func getFDCount(t *testing.T) int {
+	if runtime.GOOS != "linux" {
+		return 0
+	}
+	fds, err := os.ReadDir("/proc/self/fd")
```

Comment:
I believe tests relying on fd count of the current process would be flaky when the new or existing tests use `t.Parallel` and any of those tests also recreate any files.